src/profilers/profile-serializer.ts

```diff
@@ -0,0 +1,139 @@
+/**
+ * Copyright 2017 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {perftools} from '../profile';
+import {getIndexOrAdd} from '../util';
+import {TimeProfile, TimeProfileNode} from '../v8-types';
+// A stack of function UIDs.
+type Stack = Array<number>;
+
+/**
+ * Converts v8 Profile into profile with profile format used by Stackdriver
+ * Profiler.
+ *
+ * @param prof - profile to be converted.
+ * @param intervalMicros - average time (microseconds) between samples.
+ */
+export function serializeTimeProfile(
+    prof: TimeProfile, intervalMicros: number) {
+  let samples: Array<perftools.profiles.Sample> = [];
+  let locations: Array<perftools.profiles.Location> = [];
+  let functions: Array<perftools.profiles.Function> = [];
+  let locationMap: Map<number, perftools.profiles.Location> = new Map();
+  let functionMap: Map<number, perftools.profiles.Function> = new Map();
+  let strings = [''];
+
+  let sampleValueType = createSampleValueType();
+  let timeValueType = createTimeValueType();
+
+  serializeNode(prof.topDownRoot);
+
+  return {
+    sampleType: [sampleValueType, timeValueType],
+    sample: samples,
+    location: locations,
+    function: functions,
+    stringTable: strings,
+    // opt drop_frames
+    // opt keep_frames
+    timeNanos: 1000 * 1000 * prof.endTime,                         // Nanos
+    durationNanos: 1000 * 1000 * (prof.endTime - prof.startTime),  // Nanos
+
+    periodType: timeValueType,
+    period: intervalMicros
+  };
+
+  /**
+   * Adds samples from a node and it's children to the fields tracking
+   * profile serialization.
+   *
+   * @param node - the node which is serialized
+   * @param stack - the stack trace to the current node.
+   */
+  function serializeNode(root: TimeProfileNode) {
+    let nodes = [root];
```

Comment:
Since you always push/pop nodes and stacks together, this suggests that you you should probably define a `Entry` data structure with a node and a stack. Push and pop atomic entities of this type.